RFD 52/53/54: Passwordless

codingllama opened the pull request:
Design for passwordless/usernameless support in Teleport.

The design is split in various documents in order to keep them smaller and more focused. The current set of designs covers a fully-functioning local auth solution, including:

* Server-side changes
* Browser support
* FIDO2 CLI support (Linux and macOS)
* macOS CLI support for Touch ID

codingllama committed:
Add RFD: Passwordless

codingllama committed:
Add RFD: Passwordless for FIDO2 clients

codingllama commented:
FYI @klizhentas @russjones @alex-kovoy.

codingllama commented:
Relevant for issues #9160 and #8499.

alex-kovoy commented on rfd/9999-passwordless-fido2.md:
```diff
@@ -0,0 +1,551 @@
+There are essentially two options for CLI clients / `tsh`:
+
+1. Leverage browsers and delegate FIDO2 authentication to them; or
+2. Write CLI-native FIDO authentication.
```
I do like the option (1) with all the benefits it provides. As mentioned, it gives us a free-multi-platform support which should not be underestimated. 

As for downside, I am not sure if purely FIDO2 CLI-based flow matters that much. This is because `tsh` will always require a browser for SSO which is by far the most popular Teleport auth preference. And since we know that browser UX _wins_ over CLI (with less taps), it seems like option (1) is preferable.

And another benefit is the Teleterm (upcoming native Teleport client) will be able to use `tsh` APIs as is :)

codingllama replied:
Yep, I agree with you there. It's hard to ignore the advantages of this option.

klizhentas replied:
@codingllama @alex-kovoy what if we limit passwordless flow only if folks are using teleterm? Can this credentials handler used in teleport terminal?

alex-kovoy replied:
We can support both options (1) and (2) in `teleterm`. It's just option (2) will require minor changes (when prompting for user input) which is not a big deal.

phosphore replied:
While delegating FIDO2 authentication to browsers is attractive, won’t this “train” users to trust a browser pop-up (a thing attackers can also easily create) when starting the FIDO ceremony? Restricting this to Teleterm would be best since you would also have greater control in respect to using the default user agent for the system (through Electron it's also possible to enforce certificate pinning).

codingllama replied:
I think this ship has sailed with the various Cloud CLI tools doing browser popups, but it's a fair point.

alex-kovoy commented on rfd/9999-passwordless.md:
```diff
@@ -0,0 +1,501 @@
+Authentication looks similar to our well-known challenge/response protocols,
+with the exception that the initial challenge is created _without_ knowledge of
+the user identity. This is a departure from the current authn model where
+[challenge generation is gated on username and password](
```
> challenge is created _without_ knowledge of the user identity.

Do we need to worry about an abuse of requests to this backend public endpoint that creates challenges for anonymous users?

codingllama replied:
I talk a bit about this in the Security section. 

I think, in the end, this is an inherent problem for authentication endpoints - this one is not that different. We can (and should!) rate limit access to mitigate bad behavior, but there isn't a whole lot to do other than that. (I'm open to ideas if you have any.)

alex-kovoy replied:
I think the difference with existing public auth endpoints is that this one introduces a side affect as it creates a challenge (stored in the memory/db). So one idea would be to use a circular in-memory buffer with a reasonable size (in addition to rate limits) for storing anonymous login state.

codingllama replied:
Noted - thanks for the suggestion. I'll add a nod to in the design later.

codingllama replied:
Done.

codingllama committed:
Add RFD: Passwordless for macOS CLI

phosphore commented on rfd/9999-passwordless-fido2.md:
```diff
@@ -0,0 +1,551 @@
+       │                    │                   CreateAssigneeCredentials(sshPublicKey)                   │
+       │                    │ ────────────────────────────────────────────────────────────────────────────>
+       │                    │                                  │                                          │
+       │                    │                                  assigneeID                                 │
```
Wouldn't this facilitate MITM attacks? It seems that the `assigneeID` secret is a single point of failure for the ceremony. What is preventing an attacker from serving their own `/web/assignee?assigneeID=n` links and phish a victim?

codingllama replied:
I think there are a few protecting factors here:

1) `tsh` is the sole holder of the private key, fishing for the assignee ID won't give you that (the key never leaves tsh)
2) To MITM `/web/assignee` you need to impersonate the Teleport Proxy the user talks to. Assuming a proper prod setup (ie, no `tsh login --insecure`) this is hopefully hard to do.

 It's also important to note that the assignee ID is not really a secret, although it is hard to guess beforehand.

phosphore replied:
I'm sorry, I could have been clearer. In the theorized attack scenario:
* An adversary `tsh` starts authentication by registering a new assignee login in Teleport, including the public key to be signed.
* The victim is tricked into completing the authentication ceremony using the resulting assignee ID in their web browser, who performs the authentication ceremony on the evil `tsh`'s behalf.
* Once the ceremony is complete, the evil `tsh` retrieves the credentials from Teleport.

Is the above scenario reasonable or am I reading the design flow incorrectly/missing something?

codingllama replied:
An adversary `tsh` that manages to trick the user could run the login ceremony by itself as well, regardless of passwordless or assignee logins, but I think I get your point now.

What I think we could do to mitigate is to make it extra clear, in the Web UI, that we are running login on behalf of `tsh` and show all the relevant information (binary name, user, origin IP, so on, as captured by the Teleport servers).

Do you any ideas to change or secure the flow? This is an optional bit of the design at the moment, but I do foresee it being useful for Terminal.

phosphore replied:
I feel that we'll need to open a dedicated RFC just for this alternative, since a new threat modeling process is needed to identify the right mitigations for this flow. One of the security needs I can anticipate is the need for implementing something similar to the OAuth Proof Key for Code Exchange pattern. PKCE ([RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)) is an extension to the Authorization Code flow that would prevent certain attacks and that is needed to securely perform the OAuth exchange from public clients. The technique involves the client first creating a secret, hashing it, and then using that secret again when exchanging the authorization code for an access token. This way if the code is intercepted, it will not be useful since the token request relies on the initial secret. See https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works for a complete diagram of the flow. Depending on the implementation, `tsh` may need to implement a similar proof to get the credentials.

codingllama replied:
Thanks for the detailed reply, Lorenzo. I'll catch up with the links and do any follow up questions later.

No objections on my part for taking this section out. We can move forward without it and cross the Terminal bridge when we have to (maybe they can move to "pure" Web authentication or get away with using `tsh` flows).

codingllama replied:
Removed section from the design. We can revisit in a future design, if at all.

klizhentas replied:
@codingllama for `Teleport Terminal`, as it's an electron app, can we simply use the `web` flow using `credentials provider` without any pop ups?

phosphore commented:
Aside from rate-limiting the challenge endpoints, are [account locks](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md) disabled when Passwordless authN is enabled?

phosphore commented:
Is Passwordless auth required globally for all users / tenant, just like U2F/OTP? In this way an attacker can't enable FIDO passwordless authN first, locking the user out.

codingllama commented:
> Aside from rate-limiting the challenge endpoints, are [account locks](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md) disabled when Passwordless authN is enabled?

I expect account locks should still apply - at least I don't see why not right now. Do you have an exception in mind?

> Is Passwordless auth required globally for all users / tenant, just like it already works in the U2F/OTP design? In this way an attacker can't enable FIDO passwordless authN first, locking the user out.

The way it is designed right now, if WebAuthn is enabled then any user could register and use passwordless, but it doesn't preclude users from authentication with a "passwordfull" method.

It's also important to notice that registration requires a prior authentication ceremony, so to lock someone out in this sense an attacker would need to fully authenticate first - and at this stage the user was already fully compromised.

codingllama commented on rfd/9999-passwordless-fido2.md:
```diff
@@ -0,0 +1,551 @@
+            │                   │         *chooses*        │                                          │
+            │                   │ ─────────────────────────>                                          │
+            │                   │                          │                                          │
+            │               sign challenge                 │                                          │
```
We check the clone warnings in our current implementation, but I've elected to only log a warning (see [lib/auth/webauthn/login.go#L227](https://github.com/gravitational/teleport/blob/d3dca90b93f648f128c11c42ab7bcda61893ceec/lib/auth/webauthn/login.go#L227)). As you said, it's a bit of a toss whether sign counters are reliable and should be treated as a failure, but happy to change if you folks it's for the best.

(This plays a part in our WebAuthn implementation, but not so much for pwdless itself.)

codingllama commented on rfd/9999-passwordless.md:
```diff
@@ -0,0 +1,501 @@
+Teleport may choose to ship with canned set of allowed passwordless roots. An
+example of said roots could include:
+
+* [Apple WebAuthn Root CA](https://www.apple.com/certificateauthority/private/)
```
I'd be leaning on canned retrieval here, but I'll make a note in the design. Do you have any specific checks in mind, in any case?

phosphore replied:
For the FIDO Metadata Service:
* Follow the [metadata TOC processing rules](https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-metadata-service-v2.0-rd-20180702.html#metadata-toc-object-processing-rules) and verify the signature of the Metadata TOC object using the TOC signing certificate chain
* Check whether the status report of the authenticator model has changed compared to the cached entry (e.g. an authenticator changed from the status `REVOKED` to `FIDO_CERTIFIED` unexpectedly)
* Cross-reference the TOC returned with third party metadata (the relying party, Teleport, should obtain metadata directly from authenticator vendors or other trusted sources and cross-checks it)
* Raise a log warning if it changes from the latest cached version

For the Apple attestation statement it could be harder, but you could try to pin the TLS certificate for the endpoint if it's observed to be stable.

codingllama commented on rfd/9999-passwordless.md:
```diff
@@ -0,0 +1,501 @@
+since devices fall back to PINs. This keeps the implementation uniform and
+without hidden behaviors.
+
+(2) allows Teleport to ship with a canned set of roots, but ditches those in
```
(1) is my personal preference, but we need actual product here. @klizhentas is the person to ask.

codingllama replied:
Picked (2) / shipping with a canned set of roots, based on Sasha's feedback (see https://github.com/gravitational/teleport/pull/9296#discussion_r797232387).

The actual roots are TBD and I don't plan on specifying them in the design (that is more of a fluid product decision).

codingllama committed:
Incorporate reviewer notes in the design

* Circular buffer for challenge storage
* Warning about dynamic root retrieval

codingllama commented:
Hello folks,

I've just pushed here the first draft for [macOS Passwordless using Touch ID](https://github.com/gravitational/teleport/pull/9296/commits/81c3ba9efdf5815653c3d10cfaf81d9bd2393478).

The macOS draft completes everything necessary for local auth passwordless. At this stage we could implement a fully-functional local auth solution, so I won't add further designs to this PR (namely SSO and Windows support).

Please let me know your thoughts!

@klizhentas @russjones @phosphore

codingllama committed:
Assign RFD numbers

codingllama committed:
Mention Touch ID prompt in support commands

codingllama commented:
FYI @r0mant @zmb3

codingllama commented:
Hey @nklaassen, could I bother you for a review as well? Thanks!

codingllama committed:
Remove browser-based CLI authentication alternative

Removed due to security concerns, may warrant a design of its own in the
future.

References:

* PKCE RFC - https://datatracker.ietf.org/doc/html/rfc7636
* Auth0 PKCE documentation -
  https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works

codingllama committed:
Include detailed UX changes in the designs

codingllama commented:
I've just pushed a few commits removing the "browser-based CLI login" alternative and adding a bunch more details in the UX sections. No changes in the engineering design.

FYI @xinding33 for UX feedback.

codingllama commented:
(Btw, one of the commit names looks off, going to amend it now.)

nklaassen commented on rfd/0053-passwordless-fido2.md:
```diff
@@ -0,0 +1,494 @@
+
+The solution is similar to the one used by browsers - we make all hardware keys
+blink and ask the user to choose the desired key by touching it. Authentication
+proceeds by enumerating the credentials from that key and letting the user,
+again, pick the desired one. (See libfido2's [examples/select.c](
```
what do these credentials look like, are they readable enough that it's not completely unreasonable to make the user pick?

codingllama replied:
They have human-readable field specifically for this purpose (ie the Teleport username), so should be fine.

https://www.w3.org/TR/webauthn-2/#dictionary-user-credential-params

nklaassen commented on rfd/0053-passwordless-fido2.md:
```diff
@@ -0,0 +1,494 @@
+as the method to choose the authenticator. This technique requires one user
+touch/verification and avoids the need for PINs with biometric authenticators.
+
+(3) is when we enumerate resident keys for the {authenticator, RPID} pair. We
+can skip user interaction if only a single key exists.
```
I'm guessing there will be a unique RPID per teleport cluster? Would it ever make sense for a user to have multiple credentials on the same hardware key for Teleport? If not, is there any way for us to prevent them from doing that?

codingllama replied:
Correct, we have a unique RPID per cluster. It's a similar situation to the old U2F AppID.

They may have credentials for multiple Teleport users (probably rare for the average user, but I do it all the time). It's fine if they do, WebAuthn foresees that, plus I don't see any problems in allowing it.

nklaassen replied:
I meant, can a single Teleport user have multiple credentials on the same hardware key? IE both credentials reference the same Teleport user in the backend. I could see someone configuring that by mistake (maybe while trying to add a second/backup key) and getting confused by seeing multiple logins with the same name. Probably not a big deal though, they could just delete duplicates

codingllama replied:
Oh, got it. [We don't allow that for WebAuthn](https://github.com/gravitational/teleport/blob/bd01152781443609fad53ab52722c9c38785eb53/lib/auth/webauthn/register.go#L138-L153), so should be safe in this front.

nklaassen commented on rfd/0053-passwordless-fido2.md:
```diff
@@ -0,0 +1,494 @@
+Security considerations are largely unchanged in relation to the
+[Passwordless RFD][passwordless rfd].
+
+It's important to note that CLI applications with hardware key access are
+particularly dangerous, as they are not subject to the security measures
+we've come to expect from browsers (in fact, _any_ untrusted application that
```
Would `tsh` need any special permissions to access hardware keys on any of the supported OS? If any, can they be requested at runtime? I'm hoping we don't need to run as root on linux or with admin rights on Windows, or something like that.

ninja edit - I see this is covered in the next doc for MacOS. Do we need to do anything similar for Linux or Windows?

codingllama replied:
Nope, we don't need special permissions for FIDO2 - in fact it's the exact same thing we do today for U2F. That is why you should be wary of apps interacting with your Yubikeys - they could be pretending to be anyone.

MacOS / Touch ID is a different beast, in general it's much safer in terms of permissions and app isolation, but one should still be careful with whom they trust.

nklaassen commented on rfd/0053-passwordless-fido2.md:
```diff
@@ -0,0 +1,494 @@
+[webauthn](https://github.com/flynn/u2f/tree/webauthn) branch that could be used
+as a starting point.
+
+The alternative seems a bit extreme, at least for the moment, so it wasn't
```
```suggestion
This alternative seems a bit extreme, at least for the moment, so it wasn't
```

nklaassen reviewed (commented):
How does this interact with SAML/SSO logins, is it an alternative for local users only?

codingllama commented:
@nklaassen 

> How does this interact with SAML/SSO logins, is it an alternative for local users only?

It doesn't (for now), the designs here are local-only.

codingllama committed:
Fix spelling in 0053-passwordless-fido2.md

Co-authored-by: Nic Klaassen <nic@goteleport.com>

klizhentas commented on rfd/0054-passwordless-macos.md:
```diff
@@ -0,0 +1,424 @@
+To allow users agency over the eager behaviors of Touch ID, `tsh` is augmented
+with the following flags:
+
+* `tsh --pwdless_mode={auto,on,off}` - activate/deactivate passwordless
```
@codingllama any reason that can't be set by the `auth-preferences` by admin?

klizhentas commented on rfd/0054-passwordless-macos.md:
```diff
@@ -0,0 +1,424 @@
+
+    `off` disables passwordless logins
+
+* `tsh --mfa_mode={auto,platform,cross-platform}` - choose whether to use
```
Can we make auto choice here? For MacOS make this `TouchID` for Linux, make this `FIDO2` and remove the flag?

codingllama replied:
The reasoning for the `--mfa_mode` flag is similar to https://github.com/gravitational/teleport/pull/9296#discussion_r797730723 - `auto` locks the user on Touch ID, but they may want to use a FIDO2 key for some reason, so this exists to allow that. We could make the flag hidden, but as in the other thread, I do find it useful to have.

That said, I would expect most users to rely on the default/auto behavior and never touch this.

codingllama commented on rfd/0052-passwordless.md:
```diff
@@ -0,0 +1,516 @@
+since devices fall back to PINs. This keeps the implementation uniform and
+without hidden behaviors.
+
+(2) allows Teleport to ship with a canned set of roots, but ditches those in
```
Easy enough to do.

I'll update the design to say we are going with a canned set of roots for passwordless, but we drop those as soon as the admin defines any attestation lists.

codingllama committed:
Make passwordless ship with a canned set of allowed roots

codingllama commented on rfd/0052-passwordless.md:
```diff
@@ -0,0 +1,516 @@
+a potential source of DoS and storage space attacks, the design recommends
+adding rate-limiting for those endpoints (a possibly overdue change).
+
+A circular in-memory buffer (or similar solution) to storage challenges may also
```
Done. Spelled out mitigations more clearly in the design, including a (simple) IP rate limiter and a circular in-memory buffer for challenges.

codingllama committed:
Mention IP-based rate limiter in security mitigations

codingllama committed:
Document that `tsh` needs to be package in an app for macOS

codingllama commented:
Pushed https://github.com/gravitational/teleport/pull/9296/commits/c331f80bd6b2c3f8d9adf95e52077169348f6c99, documenting the need for a macOS .app for `tsh`. This closes the last outstanding TODO.

codingllama committed:
Spell out RPID

codingllama commented on rfd/0052-passwordless.md:
```diff
@@ -0,0 +1,518 @@
+passwordless-capable).
+
+In regards to authenticator interactions (2), it is important to note that the
+user may have multiple credentials attached to an RPID/Teleport.
```
Thanks for pointing it out, done.

zmb3 commented on rfd/0054-passwordless-macos.md:
```diff
@@ -0,0 +1,419 @@
+
+The commands are only available on macOS builds.
+
+`tsh touchid diag` - prints diagnostics about Touch ID support (for example, if
```
Great idea!

codingllama replied:
I like my debug commands. :)

This one is borne of paranoia that a client somewhere will fail under mysterious reasons, so there's at least some data we can ask them for.

codingllama commented on rfd/0053-passwordless-fido2.md:
````diff
@@ -0,0 +1,494 @@
+[CLI-native authentication](#cli-native-authentication) section):
+
+```shell
+$ tsh login --proxy=example.com --pwdless
````
For FIDO2 I think so. The best we can tell is if there are any hardware keys plugged in the system, from this point on we would have to assume that one of them holds a passwordless key and start the flow. If we guessed wrong there won't be a resident key and we'll fail partway, with I think at least one needless touch, so not that great.

Touch ID makes this a lot nicer, since in that case we can tell if a key exists without user interaction.

phosphore reviewed (approved):
With the latest additions the RFD design looks solid to me, good job @codingllama!

xinding33 reviewed (commented):
Overall looks great. Only significant comments are around the UX for `TouchID` in the case where multiple users exist for a `TouchID` credential.

codingllama commented on rfd/0053-passwordless-fido2.md:
```diff
@@ -0,0 +1,494 @@
+*taps*
+> Tap your *new* security key
+*taps*
+> Set up a new PIN for your security key:
+*enters PIN*
+> Confirm your new security key PIN:
+*enters PIN*
+> MFA device "pwdless-key" added.
```
libfido2 does that for us, so it's not that bad. If libfido2 happens to fail for a particular device, then there isn't much we can do regardless.

PIN registration flows usually have retries built-in (eg, in Chrome). I'm not planning to do that - we try PIN+confirmation once, if it fails then the user can run the command again. (Up for discussion, of course, but the simpler it is, lesser the chance I do something wrong.)

xinding33 replied:
Thanks for the clarification!

codingllama committed:
Use kebab-case for flags

codingllama commented on rfd/0054-passwordless-macos.md:
```diff
@@ -0,0 +1,419 @@
+`tsh login`, but also for any commands that require re-authentication (such as
+`tsh mfa add`).
+
+Finally, if there are Touch ID credentials for multiple users and the login user
+is not known, `tsh login` may prompt the user to specify the `--user` flag.
```
That was actually my first draft, more below.

> From a security perspective, we show the TouchID user all available users tied to their TouchID credential. That seems fine though because all these users do belong to the TouchID user.

The information leakage is what prompted me to change it to the more "obfuscated" version. I'm not concerned about the legitimate user, but if a third party that gets their hands on a (unlocked) macOS they could see which credentials it holds by doing `tsh login --proxy=example.com` and cancelling after the prompt. It's not huge, but I'd rather sidestep it if I can.

xinding33 replied:
Understood. The risk, from a UX perspective, is that the user might not recall the exact login they want to use. What are some of the easiest ways that a user can recall their login? Perhaps reverse-i-search and checking the web app?

codingllama commented on rfd/0054-passwordless-macos.md:
```diff
@@ -0,0 +1,419 @@
+login, if possible (server allows passwordless, appropriate hardware exists, one
+credential registered for "example.com")
+
+`tsh login --proxy=example.com --user=llama` behaves as above, but using a
+specific user
```
Definitely, if there is less ambiguity than all the better.  👍

xinding33 reviewed (approved):
I expect the `Touch ID` multi-user use case to be fairly uncommon and you're right in that it presents some information leakage. I'm good with this version of the RFD! Nice job, it was a fun and educational read!

codingllama commented:
Hey folks, we have Security, UX and Eng approvals; plus @klizhentas also made a pass, so I think we're good. Going to hit the merge button now. If you have any further comments let me know, I'll be happy to open a PR to address them.

Many thanks for all the thoughtful feedback, y'all.
